docs/link-reference-definition-improvements.md

```diff
@@ -0,0 +1,128 @@
+
+So as you can see, we've prioritised GitHub Pages over GitHub Web UI for the time being.
+
+Ideally, we'd like a solution that works with both, but it's not defined yet (see [[link-reference-definitions]] for more details)
+
+#### Workaround
+
+For the time being, you can use relative `[markdown links](markdown-link.md)` syntax.
+
+**Pros:**
+
+- This will work on all platforms.
+
+**Cons:**
+
+- It will break the Markdown Notes [[backlinking]] support
+- Less convenient to write
+
+### Finding certain words clutter the VS Code search results
+
+Since link reference definitions have `[//begin]` and `[//end]` guards with explanatory text that use certain words, these words (like "generate") appear in VS Code search results if you happen to search matching strings from the workspace.
+
+## Improvement Proposal
+
+Problem space in essence:
+
+- During edit-time (when modifying the markdown files in an editor)
+  - link reference definitions are needed if user uses editor extensions that don't understand wikilinks
+  - link reference definitions may be annoying since they 
+    - add content to files that the user hasn't typed in by themselves
+    - get out of date if user uses a tool that doesn't autogenerate them
+    - may clutter the search results
+- During build-time (when converting markdown to html for publishing purposes)
+  - link reference definitions are needed, if the files are published via such tools (or to such platforms) that don't understand wikilinks
+  - link reference definitions might have to be in different formats depending on the publish target (e.g. Github pages vs Github UI)
+
+The potential solution:
+
+- For edit-time
+  - Make edit-time link reference definition generation optional via user settings. They should be on by default, and generating valid markdown links with a relative path to a `.md` file.
+  - Make format of the link reference definition configurable (whether to include '.md' or not)
+  - Out of recommended extensions, currently only "markdown links" doesn't support them (?). However even its [code](https://github.com/tchayen/markdown-links/blob/master/src/parsing.ts#L25) seems to include wikilink parser, so it might just be a bug?
+- For build-time
+  - To satisfy mutually incompatible constraints between GitHub UI, VSCode UI, and GitHub Pages, we should add a pre-processing/build step for pushing to GitHub Pages. 
+    - This would be a GitHub action (or a local script, ran via foam-cli) that outputs publish-friendly markdown format for static site generators and other publishing tools
+    - This build step should be pluggable, so that other transformations could be ran during it
+  - Have publish targets defined in settings, that support both turning the link reference definitions on/off and defining their format (.md or not). Example draft (including also edit-time aspect):
+    ```typescript
+    // settings json:
+    {
+      "foam": {
+        "publish": [
```

Comment:
Is there a reason for this to be an array of single-keyed objects? For me it would feel more natural either as object `{ [target: string]: { wikiLinks, linkRefDefs } }`, or as array of `{ target, wikiLinks, linkRefDefs }[]`

Reply:
You are absolutely correct, somehow this morphed into single-key beast which doesn't make any sense. The intent was to model it according to vscode launch configurations, which uses array of objects identified by the name attribute. The reason for array is the explicit ordering, which might get lost on keyed objects depending on the execution environment. Will fix in the following commit.